Fixing build configs. Tweaking Protocol message usage. Logging. Docs.

mcottontensor opened the pull request:
## Relevant components:
- [x] Signalling library
- [x] Common library
- [x] Frontend library
- [x] Frontend UI library
- [ ] Matchmaker
- [ ] Platform scripts
- [ ] SFU

## Problem statement:
There were several issues with the build setup relating to ESM and CJS builds where in some configurations libraries would fail to resolve.
The usage of Protocol Messages was a little confusing.
Documentation on the signalling protocol was split over common and signalling.

## Solution
I have updated the build configuration so it should be possible to now build esm or cjs in isolatiion and have it all work.
I moved the usage of protocol messages out of transport and into signalling protocol.
I have moved the signalling documentation into common.
I have updated the logger usage in common so it can be overridden since the Logger in common and Logger in signalling handle things differently. It's not the best solution but its a solution for now.
I have rebuilt the docs for both common and signalling so it should be up to date.

## Documentation
Built.

mcottontensor commented:
Closes #257
Closes #256
Closes #255

lukehb commented on SFU/README.md:
```diff
@@ -0,0 +1,59 @@
+| Name | Type | Default | Description |
+|-|-|-|-|
+| rtcMinPort | Number | 40000 | Minimun RTC port for ICE, DTLS, RTP, etc. |
+| rtcMazPort | Number | 49999 | Maximum RTC port for ICE, DTLS, RTP, etc. |
```
Should be `rtcMaxPort`

lukehb commented on SFU/README.md:
```diff
@@ -0,0 +1,59 @@
+| Name | Type | Default | Description |
+|-|-|-|-|
+| signallingURL | String | 'http://localhost:8889' | The URL pointing to the signalling server we want to connect to. |
+| SFUId | String | 'SFU' | The streamer name this peer will be given when connecting to the signalling server. Peers wishing to receive from this SFU should subscribe to this ID. |
```
I think 

> The ~~streamer~~ name this peer will be given when connecting to the signalling server. Peers wishing to receive from this SFU should subscribe to this ID.

lukehb commented on SFU/README.md:
```diff
@@ -0,0 +1,59 @@
+
+Launch the streaming app with the following arguments
+`-SimulcastParameters="1.0,5000000,20000000,2.0,1000000,5000000,4.0,50000,1000000"`
+This tells the pixel streaming plugin to stream simulcast with 3 streams, each one scaling down by 2. The sequence of values is as follows, `scale_down_factor,min_bitrate,max_bitrate,...repeating for each stream`
```
I prefer:

> scaling video resolution by half.

lukehb replied:
> pixel streaming

to 

> Pixel Streaming

lukehb reviewed (changes requested):
Minor changes requested.

DenisTensorWorks commented on SFU/README.md:
```diff
@@ -0,0 +1,59 @@
+`-SimulcastParameters="1.0,5000000,20000000,2.0,1000000,5000000,4.0,50000,1000000"`
+This tells the pixel streaming plugin to stream simulcast with 3 streams, each one scaling down by 2. The sequence of values is as follows, `scale_down_factor,min_bitrate,max_bitrate,...repeating for each stream`
+
+When this streams to the SFU the SFU will detect these 3 streams and then selectively stream these out to connected peers based on their connection quality.
```
> When this streams to the SFU the SFU will detect

When this streams to the SFU, the SFU will detect

DenisTensorWorks commented on SignallingWebServer/README.md:
```diff
@@ -0,0 +1,32 @@
+# Pixel Streaming Signalling Server
+
+The signalling server is a small middle-man application that sits between streamers and other peers. It handles the initial connection negotiations and some other small ongoing control messages between peers as well as acting as a simple web server for serving the [Frontend](/Frontend/README.md) web application.
```
> a small middle-man application

a small intermediary application

DenisTensorWorks commented on SignallingWebServer/README.md:
```diff
@@ -0,0 +1,32 @@
+
+## Configuration
+
+Configuration of the signalling server is handled via the config.js file in the SignallingWebServer directory. The following are supported options.
```
> The following are supported options.

The following are its supported options.

DenisTensorWorks commented on SignallingWebServer/README.md:
```diff
@@ -0,0 +1,32 @@
+| HomepageFile | String | 'player.html' | The root file of the frontend web application. |
+| AdditionalRoutes | Map | | Additional routes for the web application. |
+| EnableWebserver | Boolean | true | Enables or disables the serving of the frontend through the internal web server. Disbable this if you are serving your own frontend. |
+| MatchmakerAddress | String | | The IP/hostname of the match maker application. |
```
> the match maker application

the matchmaker application

DenisTensorWorks commented on SignallingWebServer/README.md:
```diff
@@ -0,0 +1,32 @@
+| AdditionalRoutes | Map | | Additional routes for the web application. |
+| EnableWebserver | Boolean | true | Enables or disables the serving of the frontend through the internal web server. Disbable this if you are serving your own frontend. |
+| MatchmakerAddress | String | | The IP/hostname of the match maker application. |
+| MatchmakerPort | Number | 9999 | The port the match maker is listening on. |
```
> the match maker is listening on

the matchmaker is listening on

DenisTensorWorks commented on SignallingWebServer/README.md:
```diff
@@ -0,0 +1,32 @@
+| EnableWebserver | Boolean | true | Enables or disables the serving of the frontend through the internal web server. Disbable this if you are serving your own frontend. |
+| MatchmakerAddress | String | | The IP/hostname of the match maker application. |
+| MatchmakerPort | Number | 9999 | The port the match maker is listening on. |
+| PublicIp | String | "localhost" | The public IP/hostname of the host that the signalling server is listening on. This is used by the match maker. |
```
> This is used by the match maker.

This is used by the matchmaker.

DenisTensorWorks reviewed (commented):
Minor style and sentence flow edits.

mcottontensor committed:
- Fixed issue with esm and cjs builds not being correctly configured
- Adjusting how Protocol and Transport handle protocol BaseMessages. Transport should just handle strings. Protocol should handle messages.

mcottontensor committed:
Fixing up logging from Common in signalling lib

mcottontensor committed:
Moved signalling protocol documentation

mcottontensor committed:
Rebuilt common docs

mcottontensor committed:
Rebuilt signalling docs

mcottontensor committed:
Frontend watch should use dev

lukehb commented on Frontend/ui-library/tsconfig.base.json:
```diff
@@ -10,7 +10,9 @@
         "noFallthroughCasesInSwitch": true,
         "noImplicitReturns": true,
-        "noPropertyAccessFromIndexSignature": true
+        "noPropertyAccessFromIndexSignature": true,
```
Interesting, was there issues without this?

mcottontensor replied:
I didn't change it, just added the comma.

lukehb commented on package.json:
```diff
@@ -18,6 +18,8 @@
     "private": true,
     "scripts": {
         "build": "npm run build --ws",
+        "build:all:cjs": "cd Common && npm run build:cjs && cd ../Signalling && npm run build:cjs && cd ../SignallingWebServer && npm run build && cd ../Frontend/library && npm run build:cjs && cd ../ui-library && npm run build:cjs && cd ../implementations/typescript && npm run build:dev",
```
Another option could be `npm run build:cjs --ws` as long as every package has that option (even if is a no-op in `Extras/`). I think this is probably better for specificity though and order of operations.

mcottontensor replied:
Well the order is important. I assume the --ws just follows the order in the workspace entries but thats kind of loose. We also have some other packages in the repo that arent required for development.

lukehb replied:
Agreed

lukehb reviewed (approved):
Approved, left a few comments and questions. But no changes required, looks good to merge,
